animation_nodes/nodes/gpencil/gp_layer_info.py

```diff
@@ -0,0 +1,136 @@
+        elif self.frameType == "INDEX":
+            if self.useFrameList:
+                return "execute_FrameIndices"
+            else:
+                return "execute_FrameIndex"
+        elif self.frameType == "FRAME":
+            if self.useFrameList:
+                return "execute_FrameNumbers"
+            else:
+                return "execute_FrameNumber"
+        elif self.frameType == "ALL":
+            return "execute_AllFrames"
+
+    def execute_ActiveFrame(self, layer, scene):
+        frames = layer.frames
+        if len(layer.frames) == 0:
+            return GPFrame(), layer.layerName, layer.blendMode, layer.opacity, layer.passIndex
+
+        return (self.getActiveFrame(scene.frame_current, frames), layer.layerName, layer.blendMode,
+                layer.opacity, layer.passIndex)
+
+    def execute_FrameIndex(self, layer, frameIndex):
+        frames = layer.frames
+        if len(layer.frames) == 0:
+            return GPFrame(), layer.layerName, layer.blendMode, layer.opacity, layer.passIndex
+
+        frame = self.getFrame(frames, frameIndex)
+        return frame, layer.layerName, layer.blendMode, layer.opacity, layer.passIndex
+
+    def execute_FrameIndices(self, layer, frameIndices):
+        frames = layer.frames
+        if len(layer.frames) == 0:
+            return [], layer.layerName, layer.blendMode, layer.opacity, layer.passIndex
+
+        outFrames = [self.getFrame(frames, index) for index in frameIndices]
+        return outFrames, layer.layerName, layer.blendMode, layer.opacity, layer.passIndex
+
+    def execute_FrameNumber(self, layer, frameNumber):
+        frames = layer.frames
+        if len(layer.frames) == 0:
+            return GPFrame(), layer.layerName, layer.blendMode, layer.opacity, layer.passIndex
+
+        frameNumbers = self.getFrameNumbers(frames)
+        index = self.getFrameNumberIndex(frameNumbers, frameNumber)
+        return frames[index], layer.layerName, layer.blendMode, layer.opacity, layer.passIndex
+
+    def execute_FrameNumbers(self, layer, inFrameNumbers):
+        frames = layer.frames
+        if len(frames) == 0:
+            return [], layer.layerName, layer.blendMode, layer.opacity, layer.passIndex
+
+        frameNumbers = self.getFrameNumbers(frames)
+        outFrames = [frames[self.getFrameNumberIndex(frameNumbers, frameNumber)] for frameNumber in inFrameNumbers]
+        return outFrames, layer.layerName, layer.blendMode, layer.opacity, layer.passIndex
+
+    def execute_AllFrames(self, layer):
+        frames = layer.frames
+        return frames, layer.layerName, layer.blendMode, layer.opacity, layer.passIndex
+
+    def getActiveFrame(self, currentFrame, frames):
+        frameNumbers = self.getFrameNumbers(frames)
+        if currentFrame <= frameNumbers.getMinValue(): return frames[0]
+        allFrames = list(filter(lambda x: x <= currentFrame, frameNumbers))
+        return frames[frameNumbers.index(allFrames[-1])]
```

Comment:
I think this is simpler and faster.
```python
min(frame for frame in frames if frame.frameNumber >= currentFrame, key = lambda f : f.frameNumber)
```

Reply:
Omar this expression gave me an error. Can you please explain briefly what is going on in this line?

Reply:
Sorry, my bad. We want to find the frame with the largest frame number that is less than or equal the current frame. So we first filter the frames whose frame number is less than or equal the current frame, we do that using the generator:

```python
(frame for frame in frames if frame.frameNumber <= currentFrame)
```

To get the frame which has maximum frame number, we just use max with the `frameNumber` as key. So the expression is: 

```python
max((frame for frame in frames if frame.frameNumber <= currentFrame), key = lambda f : f.frameNumber)
```

Reply:
Omar, the above code works fine when `currentFrame >= frames[0].frameNumber`, otherwise it gives error because sequese(in max function) becomes empty. Here is workaround,
```
def getActiveFrame(self, currentFrame, frames):
        frameNumbers = self.getFrameNumbers(frames)
        if currentFrame <= min(frameNumbers): return frames[0]
        return max((frame for frame in frames if frame.frameNumber <= currentFrame),
                    key = lambda frame : frame.frameNumber)
```
or
```
def getActiveFrame(self, currentFrame, frames):
        try: return max((frame for frame in frames if frame.frameNumber <= currentFrame),
                    key = lambda frame : frame.frameNumber)
        except: frames[0]
```

Reply:
The max function can take a default value, I think.
https://docs.python.org/3/library/functions.html#max

Reply:
Thanks, it worked.